zilliqa/src/exec.rs

```diff
@@ -826,14 +890,183 @@ impl PendingState {
+            StorageValue::Value(value) => Ok(value),
+        }
+    }
+
+    pub fn load_storage_by_prefix(
+        &mut self,
+        address: Address,
+        var_name: &str,
+        indices: &[Vec<u8>],
+    ) -> Result<BTreeMap<Vec<u8>, StorageValue>> {
+        let account = load_account(&self.pre_state, &mut self.new_state, address)?;
+
+        // Even if we have something cached for this prefix, we don't know if it is a full representation of the map.
+        // It might have been the case that only a few subfields were cached. Therefore, we need to retrieve the full
+        // map from disk and apply any cached (and potentially updated) values. In future, we should use the 'complete'
+        // flag on maps to avoid needing to read from disk unconditionally.
+
+        let values_from_disk: Vec<_> = self
+            .pre_state
+            .get_account_trie(address)?
+            .iter_by_prefix(&storage_key(var_name, indices))?
+            .collect();
+
+        let mut map = StorageValue::complete();
+        let cached = account.storage.get(var_name);
+
+        // Un-flatten the values from disk into their true representation.
+        for (k, v) in values_from_disk {
+            let (disk_var_name, disk_indices) = split_storage_key(&k)?;
+            assert_eq!(var_name, disk_var_name);
+            assert!(disk_indices.starts_with(indices));
+
+            let mut current_value = &mut map;
+            let mut current_cached = cached;
+
+            for index in disk_indices {
+                if let Some(c) = current_cached {
+                    match c {
+                        StorageValue::Map { map, .. } => {
+                            current_cached = map.get(&index);
+                        }
+                        // This branch can be hit in two cases:
+                        // * Firstly, we expect the final index to point to a value in the cache, rather than a map.
+                        // * Secondly, if a portion of the map has been deleted, then the cache can contain a `None`
+                        // value at a greater height than the depth of the map.
+                        StorageValue::Value(_) => {}
+                    }
+                }
+
+                let map = match current_value {
+                    StorageValue::Map { map, .. } => map,
+                    StorageValue::Value(_) => {
+                        return Err(anyhow!("expected map"));
+                    }
+                };
+                // Note that we insert 'complete' maps here, because we know we are going to add *ALL* values from
+                // within this map.
+                current_value = map.entry(index).or_insert_with(StorageValue::complete);
             }
+
+            *current_value = if let Some(cached) = current_cached {
+                cached.clone()
+            } else {
+                StorageValue::Value(Some(v.into()))
+            };
         }
+
```

Comment:
If I:

 - Create foo.bar..baz = {"wombats": 2 }
 - Set foo.bar = "fish" (deleting the map)
 - Ask for foo.bar.baz

I think what'll happen here is that we'll walk the tree, note the value at foo.bar, ignore it and return { "wombats": 2}?

Reply:
I'm 99% sure your second bullet point is impossible. Scilla won't let you assign a value above the bottom level of a map. In other words, the type of `foo` is `Map String (Map String Int)`. When you do `foo[bar] = "fish"`, the Scilla type checker will say "expected a `(Map String Int)` but got the string `"fish"`"?

Reply:
Scilla may not, but your interface allows it and one day some bright spark (possibly me :-))  may decide to  let the Scilla interpreter allow it somehow, on which day you probably want a better way of reacting than generating an obscure bug that occasionally returns previously deleted maps - or you could prevent the situation from arising by rejecting updates to a non-terminal key at your level as well..

Reply:
Will add an assertion in a follow-up :+1: